docs/codeql/codeql-language-guides/customizing-library-models-for-go.rst

```diff
@@ -0,0 +1,413 @@
+
+
+Since we are adding a neutral model, we need to add tuples to the ``neutralModel`` extensible predicate.
+The first four values identify the callable (in this case the getter of the ``Now`` property) to be modeled as a neutral, the fifth value is the kind, and the sixth value is the provenance (origin) of the neutral.
+
+- The first value ``System`` is the namespace name.
+- The second value ``DateTime`` is the class (type) name.
+- The third value ``get_Now`` is the method name. Getter and setter methods are named ``get_<name>`` and ``set_<name>`` respectively.
+- The fourth value ``()`` is the method input type signature.
+- The fifth value ``summary`` is the kind of the neutral.
+- The sixth value ``manual`` is the provenance of the neutral.
+
+Example: Accessing the ``Body`` field of an HTTP request
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This example shows how we can model a field as a source of tainted data.
+
+.. code-block:: go
+
+   func TaintFlow(w http.ResponseWriter, r *http.Request) {
+       body := r.Body // The Body field of an HTTP request is a source of tainted data.
+       ...
+   }
+
+We need to add a tuple to the ``sourceModel``\(namespace, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+  extensions:
+  - addsTo:
+      pack: codeql/go-all
+      extensible: sourceModel
+    data:
+      - ["net/http", "Request", True, "Body", "", "", "", "remote", "manual"]
+
+Since we are adding a new source, we need to add a tuple to the ``sourceModel`` extensible predicate.
+The first five values identify the field to be modeled as a source.
+
+- The first value ``net/http`` is the package name.
+- The second value ``Request`` is the name of the type that the field is associated with.
+- The third value ``True`` is a flag that indicates whether or not the source also applies to all overrides of the field.
+- The fourth value ``Body`` is the field name.
+- The fifth value ``""`` is blank since it is a field access and field accesses do not have method signatures in Go.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the source.
+
+- The seventh value ``""`` is left blank. Leaving the access path of a source model blank indicates that it is a field access.
+- The eighth value ``remote`` is the source kind. This indicates that the source is a remote source of untrusted data.
+- The ninth value ``manual`` is the provenance of the source, which is used to identify the origin of the source.
+
+Package grouping
+~~~~~~~~~~~~~~~~
+
+Since Go uses URLs for package identifiers, it is possible for packages to be imported with different paths. For example, the ``glog`` package can be imported using both the ``github.com/golang/glog`` and ``gopkg.in/glog`` paths.
```

Comment:
Before go modules, this was much more true. Now the go.mod file specifies the URL that this module expects to be found at, which creates a canonical import path for a module (and hence the packages it contains). The reasons why we want to group models for several different import paths are:
* forks (e.g. klog was forked from glog by kubernetes);
* when a package moves, we still want to analyse code written for the old import path.
We don't have to explain this all, but I mention it so you have it in mind when writing the sentence above.